src/metabase/lib/schema/expression.cljc

```diff
@@ -1,39 +1,181 @@
+(mr/def ::*.integer
+  [:schema
+   [:cat
+    [:= :*]
+    ::common/options
+    [:* [:schema [:ref ::integer]]]]])
 
 ;;; An expression that returns an integer.
 (mr/def ::integer
-  ::literal/integer)
+  [:or
+   {:error/message "expression returning an integer"}
+   ::literal/integer
+   ::length
+   [:ref ::*.integer]
+   (ref-of-base-type :type/Integer)
+   ;; `:temporal-extract` units like `:day-of-month` actually return `:type/Integer` rather than a temporal type
+   (ref-of-base-type-with-temporal-unit :type/Date     ::temporal-bucketing/unit.date.extract)
+   (ref-of-base-type-with-temporal-unit :type/Time     ::temporal-bucketing/unit.time.extract)
+   (ref-of-base-type-with-temporal-unit :type/DateTime ::temporal-bucketing/unit.date-time.extract)])
 
-;;; An expression that returns a floating-point number.
-(mr/def ::floating-point
-  ::literal/floating-point)
+;;; a `:*` clause that isn't an `::*.integer` (i.e., one or more clauses sub-expressions is a non-integer-real expression)
+;;; returns a non-integer-real number
+(mr/def ::*.non-integer-real
+  [:and
+   [:schema
+    [:cat
+     [:= :*]
+     ::common/options
+     [:* [:schema [:ref ::number]]]]]
+   [:not ::*.integer]])
+
+;;; An expression that returns a number that includes a non-integer-real place, including but not limited floats, doubles,
+;;; BigDecimals, SmallDecimals, MediumDecimals, etc! Basically any normal number that isn't an integer!
+(mr/def ::non-integer-real
+  [:or
+   {:error/message "expression returning a number with a non-integer-real place"}
+   [:ref ::literal/floating-point]
+   [:ref ::*.non-integer-real]
+   ;; for some crazy reason `:type/Float` is the common base type of all numbers with non-integer-real places, not something
+   ;; smart like `:type/Decimal`
+   (ref-of-base-type :type/Float)])
 
 ;;; Any expression that returns any kind of number.
 (mr/def ::number
   [:or
+   {:error/message "expression returning a number"}
    ::integer
-   ::floating-point])
+   ::non-integer-real])
+
+;;; `:datetime-add` with a Date expression:
+;;;
+;;;    [:datetime-add <options> <date-expression> <amount> <date-interval-unit>]
+(mr/def ::datetime-add.date
+  [:schema
+   [:tuple
+    [:= :datetime-add]
+    ::common/options
+    #_expression [:ref ::date]
+    #_amount     [:ref ::integer]
+    #_unit       ::temporal-bucketing/unit.date.interval]])
+
+;;; Any expression that returns a `:type/Date`
+(mr/def ::date
+  [:or
+   {:error/message "expression returning a date"}
+   [:ref ::literal/date]
+   [:ref ::datetime-add.date]
+   (ref-of-base-type-with-temporal-unit :type/Date [:not ::temporal-bucketing/unit.date.extract])
+   ;; TODO -- does a truncation of a :type/DateTime to `::temporal-bucketing/unit.date.truncate` like `:day` return a
+   ;; `:type/Date`? Should we act like it does?
+   #_(ref-of-base-type-with-temporal-unit :type/DateTime ::temporal-bucketing/unit.date.truncate)])
```

Comment:
Feels like "yes, but" to me. It could also be viewed as a `:type/DateTime` that happens to be set for midnight UTC, so what we probably want is a type that can be used as either depending on the context. (`derive`?)